Implement modding support

mageven opened the pull request:
# Implement modding support

This PR adds full support for mods. It follows [Atmosphère's](https://github.com/Atmosphere-NX/Atmosphere) structure with some enhancements for host FS and emulators convenience.

# At a glance
```
// named file replacers [RECOMMENDED]
@mods/contents/<program id>/<mod dir>/romfs/*   [loose files]
@mods/contents/<program id>/<mod dir>/exefs/*   [loose files]

// named patches [RECOMMENDED]
@mods/exefs_patches/<mod dir>/<build id>.ips|*.pchtxt
@mods/nro_patches  /<mod dir>/<build id>.ips|*.pchtxt
@mods/kip_patches  /<mod dir>/<KIP SHA256>.ips

// named title-specific patches [ENHANCEMENT]
@mods/contents/<program id>/<mod dir>/exefs/<build id>.ips|*.pchtxt

// global replacers [COMPATIBILITY, with ams]
@mods/contents/<program id>/romfs.bin
@mods/contents/<program id>/exefs.nsp
@mods/contents/<program id>/romfs/*   [loose files]
@mods/contents/<program id>/exefs/*   [loose files]
```

# Usage Details

Currently `@mods` is set to use `@RyujinxAppData/mods` as base mods path. You can directly access the title mods directory from the right-click context menu in the GUI.

Title-based mods (romfs and exefs files) go into `contents/` with specific dirs. Title names are pretty flexible. As long as the first 16 characters match a unique title-id, mods are picked up from it.

Patches can be placed in per-title dirs in `contents/` or in the more simpler `*_patches` dirs shown [At a glance](#at-a-glance).

All applicable mods are searched/loaded only when an Application is started for efficiency.

With a little more code in the future (GUI), mods can be searched from additional directories.

# Feature Set

1. ExeFS Partition Replacement
2. ExeFS File Replacement
3. ExeFS Patching
4. RomFS Partition Replacement
5. RomFS File Replacement

### 1. ExeFS Partition Replacement

If an `exefs.nsp` is located for an Application, it is used instead of base ExeFS.

### 2. ExeFS File Replacement

Allows replacing the standard NSO files in an Application's ExeFS. For eg., `sdk`.
Stubbing is also supported. Just create an empty file like so- `sdk.stub`.

Current limitations: Does not replace `main.npdm`.

### 3. ExeFS Patching

Patch Dirs can contain any number of `.ips` (Supports IPS and IPS32) and `.pchtxt` (IPSwitch) patches.

It's sometimes easier to dump all patches into `exefs_patches` as you don't have to worry about Title IDs. The mod loader will select the applicable patches automatically.

**IPS/IPS32:** File names should be like so: `<nso_build_id>[.anything_you_want].ips`

**IPSwitch:** Files must have `.pchtxt` extension and have an `@nsobid-` header as the first line.

`nro_patches/` and `kip_patches/` follow `exefs_patches`. KIP patches, while searched, don't apply yet (as KIPs aren't used).

### More about ExeFS

For more info on exefs features, see Atmosphere's [loader docs](https://github.com/Atmosphere-NX/Atmosphere/blob/master/docs/components/modules/loader.md).

### 4 & 5. RomFS Patching

RomFS mods are either loose files placed in `romfs/` or inside a single container `romfs.bin`. 

The files are replaced in a layered fashion. Loose files are prioritized over any files inside `romfs.bin` which in return, are prioritized over the files in the base RomFS of the Application (same as Atmosphere). See [At a glance](#at-a-glance)

As an aside, to create a `romfs.bin` file one can use [`hactoolnet`](https://github.com/Thealexbarney/LibHac/releases). Just pass the romfs directory as input and it should create a storage file. Here's an example invocation:

```
hactoolnet.exe -t romfsbuild --outfile romfs.bin mymod/romfs
```

## Example Structure
```
Ryujinx
    - bis
    - games
    ...
    + mods
        + contents
            + 0100E95012348000 (My game name)
                + My combined mod
                    + exefs
                        = 123456789ABCDEF...ips
                        = main.stub
                    + romfs
                        = <*> [loose game files]
                + My resolution mod
                    - exefs
                + My romfs mod
                    - romfs
                ...
            + 01006a805678e000
                = exefs.nsp
                = romfs.bin
            - 01006A585fF8E000-some-name-here
            ...
        + exefs_patches
            + my game patch
                = 123456789ABCDEF...ips
                = some-test-patch.pchtxt
            - patch for another game
            ...
        - nro_patches [same as exefs_patches]
        - kip_patches [same as exefs_patches]

```

## Sample Logs
```
00:00:00.412 | Application LoadApplication: Loading as XCI.
00:00:00.449 | ModLoader QueryContentsDir: Searching mods for Title 0100E95012348000
00:00:00.451 | ModLoader QueryTitleDir: Found mod 'force-max-resolution' [E]
00:00:00.451 | ModLoader QueryTitleDir: Found mod 'hq-textures' [R]
00:00:00.452 | ModLoader QueryPatchDirs: Found NSO patch 'my-game-mod-no-outlines'
00:00:00.456 | ModLoader ApplyRomFsMods: Applying RomFS mods for Title 0100E95012348000
00:00:00.466 | ModLoader ApplyRomFsMods: Replaced 45 file(s) over 2 mod(s). Processing base storage...
00:00:00.715 | ModLoader ApplyRomFsMods: Building new RomFS...
00:00:00.737 | ModLoader ApplyRomFsMods: Using modded RomFS
00:00:01.273 | ModLoader ApplyProgramPatches: Matching IPSwitch patch '1.0.0.pchtxt' in 'force-max-resolution' bid=C2C016016115BE3D616281E67E59A7918DAFEDBB
00:00:01.274 | ModLoader Parse: IPSwitch:    # Mandatory Comment
00:00:01.274 | ModLoader Parse: IPSwitch:    print_values 0x7e767e <= 1F2003D5
00:00:01.317 | Ptc LoadExeFs: Detected exefs modifications. PPTC disabled.
00:00:01.319 | Ptc Initialize: Initializing Profiled Persistent Translation Cache (enabled: False).
```

# Dev Details

IPS based on Atmosphere's implementation.<br/>
IPSwitch, my own, from the readme description and some samples.

NRO Patching is untested with actual patches (couldn't find any).

`romfs.bin` can be useful in cases where there's a ton of files to patch. Currently when a romfs mod is loaded, all the files in it will have open handles until emulation has ended. To reduce this number to one, a single-file approach is useful. This also bypasses the OS's filesystem routines and may reduce overhead.

While LayeredFS is more accurate, it was slow to build in my tests, so currently a single RomFs instance is maintained where files are added/replaced. We can revisit adding "true" LayeredFS at a later date. They don't seem as useful for now. I'm open to suggestions.

Tangentially related to modding, there's some under-the-hood changes:

- All mod logs have their own LogClass `ModLoader`

- Executables have been re-written to use a contiguous block of memory (to make proper patching possible) and Spans are used to access segments.

- ExeFs, Npdm, ControlData and SaveData calls have been reordered as discussed with @gdkchan.

**Note:** PPTC is automatically disabled when valid changes to exefs are detected (romfs mods are still ok). It's possible for PPTC to work with mods, but the code changes currently required for a proper solution (changes to JIT and PTC) are beyond the scope of this PR. Current warning message is shown in [sample logs](#sample-logs).

## Things for the future

- Reimplement IPSwitch patches with new spec
- Add cheats support
- Working PPTC with exefs mods
- A potential GUI

## Special Thanks
Thanks to all of Ryujinx team for providing so many valuable suggestions and knowledge. And thanks to all the testers too.

mageven commented:
New commit and rebase to adjust for LibHac 0.11.2 changes to `KipReader`.
No new features added yet.

mageven commented:
Now supports exefs file replacement (with stubbing). Check sections marked NEW in PR description.
No changes to structure yet. That requires larger changes and a final decision.

mageven commented:
No external changes. Internals mostly rewritten to support Auto-scan _and_ custom locations per mod type. Will explain in detail later.

AcK77 commented on Ryujinx.HLE/HOS/ApplicationLoader.cs:
```diff
@@ -366,23 +380,23 @@ void LoadNso(string filename)
+            _fileSystem.ModLoader.ApplyExefsReplacements(TitleId, nsos);
+
+            var programs = nsos.ToArray();
+
```
Nit: remove extra empty line

AcK77 commented on Ryujinx.HLE/HOS/ModLoader.cs:
```diff
@@ -0,0 +1,427 @@
+                if (type == ModType.TitleMod && titleId == ulong.MaxValue)
+                {
+                    Logger.PrintWarning(LogClass.Application, $"Orphaned {type} without TitleId '{dir.Name}'");
+                    return null;
```
Add new empty line before a return.

AcK77 commented on Ryujinx.HLE/HOS/ModLoader.cs:
```diff
@@ -0,0 +1,427 @@
+                }
+
+                Mod m = new Mod(dir, type, titleId, enabled);
+
```
Remove extra empty line

AcK77 commented on Ryujinx.HLE/HOS/ModLoader.cs:
```diff
@@ -0,0 +1,427 @@
+                if (!check)
+                {
+                    Logger.PrintWarning(LogClass.Application, $"Invalid/Empty {type} '{m.Dir.Name}'");
+                    return null;
```
Add new empty line before a return.

AcK77 commented on Ryujinx.HLE/HOS/ModLoader.cs:
```diff
@@ -0,0 +1,427 @@
+            if (count == 0)
+            {
+                Logger.PrintInfo(LogClass.Loader, "Using base RomFS");
+                return baseStorage;
```
Add new empty line before a return.

AcK77 commented on Ryujinx.HLE/HOS/ModLoader.cs:
```diff
@@ -0,0 +1,427 @@
+            }
+
+            var exefsDirs = titleMods.Where(mod => mod.Check() && mod.Exefs.Exists)
+                            .Select(mod => mod.Exefs);
```
Alignment is a bit weird here

AcK77 commented on Ryujinx.HLE/HOS/ModLoader.cs:
```diff
@@ -0,0 +1,427 @@
+        internal void ApplyNsoPatches(ulong titleId, params IExecutable[] programs)
+        {
+            var nsoMods = NsoMods.Values
+                          .Concat(TitleMods.TryGetValue(titleId, out var titleMods) ? titleMods : Enumerable.Empty<Mod>())
```
Alignment is a bit weird here (same for the next line)

AcK77 commented on Ryujinx.HLE/HOS/ProgramLoader.cs:
```diff
@@ -44,27 +44,27 @@ public static bool LoadKip(KernelContext context, KipExecutable kip)
-                kip.Header.ProcessCategory,
-                kip.Header.TitleId,
+                kip.Name.ToString(),
+                kip.Version, // TODO: Check if category was just misnamed
```
Yeah it seems to be misnamed since the beginning, maybe we should update `ProcessCreationInfo` too

AcK77 commented on Ryujinx.HLE/Loaders/Mods/IPSPatcher.cs:
```diff
@@ -0,0 +1,117 @@
+
+                int patchSize = buf[0] << 8 | buf[1];
+
+                if (patchSize == 0)                     // RLE/Fill mode
```
```suggestion
                if (patchSize == 0) // RLE/Fill mode
```

AcK77 commented on Ryujinx.HLE/Loaders/Mods/IPSPatcher.cs:
```diff
@@ -0,0 +1,117 @@
+
+                    patches.AddFill((uint)patchOffset, fillLength, buf[0]);
+                }
+                else                                    // Copy mode
```
```suggestion
                else // Copy mode
```

3096 commented:
I actually am in the middle of an IPSwitch rewrite and updated pchtxt specs to allow some more flexibility (like multiple nso per patch text and nro support). Would be great if you could take a look at the new specs: https://github.com/3096/ipswitch/blob/gui/example/README.md
I also separated the parsing into a library (backwards compatible) (WIP): https://github.com/3096/libpchtxt
But since it's C++, it probably won't be used here. It could still be used as a reference implementation tho.

mageven commented:
Thank you for taking an interest in this PR.

Although I'm not much of a modder, from a quick read, the new spec definitely has some very nice improvements. I'll wait on experienced modders to comment on this further.

How do you plan to address backwards compatibility with this new spec? (There seem to be some breaking changes and the extension is still the same)

3096 commented:
> How stable is the new spec?
I mean, "stability" depends on the implementation. The reference implementation seemed stable, but as I still haven't implemented some of the new features in the IPSwitch homebrew rewrite, there hasn't been very extensive testing on all the aspects yet.

> Do you plan on changing it more barring any suggestions?
I think the general structure would be set unless something significant requires changing it. Apart from that, I don't see it changing too much. New tags/flags might get added if it allows additional useful features. And I am open to suggestions.

mageven commented:
Finalized the mods directory structure. The only difference is (for previous testers, sorry for the slight inconvenience)-

All Title mods are now inside a directory called `contents/`. It also completely matches Atmosphere mods layout now.

Modularized a lot of ModLoader code (easier modifications in the future) and removed many inconsistencies in the old approach. Behaviour should match Atmosphere now for many cases, internally.

I updated the PR description completely. Please have another read.

Thog commented on Ryujinx.HLE/FileSystem/VirtualFileSystem.cs:
```diff
@@ -17,11 +17,12 @@ public class VirtualFileSystem : IDisposable
         public static string SystemNandPath = Path.Combine(NandPath, "system");
         public static string UserNandPath   = Path.Combine(NandPath, "user");
-
+        
```
remove those spaces

Thog commented on Ryujinx.HLE/HOS/ApplicationLoader.cs:
```diff
@@ -31,7 +31,7 @@ public class ApplicationLoader
         private readonly VirtualFileSystem _fileSystem;
 
-        public BlitStruct<ApplicationControlProperty> ControlData { get; set; }
+        public BlitStruct<ApplicationControlProperty> ControlData { get; set; }        
```
remove those spaces at the end

Thealexbarney commented on Ryujinx.HLE/HOS/ModLoader.cs:
```diff
@@ -0,0 +1,537 @@
+using LibHac.FsSystem;
+using LibHac.FsSystem.RomFs;
+using Ryujinx.Common.Logging;
+using Ryujinx.HLE.FileSystem;
```
Using not needed

mageven commented on Ryujinx.HLE/HOS/Horizon.cs:
```diff
@@ -184,11 +184,11 @@ public Horizon(Switch device, ContentManager contentManager)
+        public void LoadKip(string kipPath)
         {
-            using IStorage fs = new LocalStorage(kipFile, FileAccess.Read);
+            using IFile kipFile = new LocalFile(kipPath, OpenMode.Read);
```
Oops. Thanks for catching. Missed it when I rebased to Thog's Kip changes.
